<div align="center">

# 1º PROJETO

| Disciplina | Semestre | Docente | Horário |
| :---: | :---: | :---: | :---: |
| PAM0466 - SISTEMAS INTELIGENTES | 2026.1 | PEDRO THIAGO VALÉRIO DE SOUZA | 3M23 4M45 |

| Discente | Matrícula |
| :---: | :---: |
| ELTON CAIO VIEIRA DE LIMA | 2020010673 |
| LUCAS VIERES ARAÚJO FARIAS | 2025022531 |
</div>

<div align="center">

## Atividades
</div>

1. Calcule a correlacão entre as variáveis independentes (AT, V, AP, RH) e a variável dependente (PE). Quais variáveis parecem ter maior influência na producão de energia?

In [1]:
import pandas

dataset = pandas.read_csv("./datasets/Power Plant Data.csv")

x = dataset[["AT", "V", "AP", "RH"]]
y = dataset["PE"]

corr = x.corrwith(y)
print(corr)

AT   -0.948128
V    -0.869780
AP    0.518429
RH    0.389794
dtype: float64


Pela interpretação dos valores, as variáveis independentes AT e V apresentam uma correlação muito forte (de magnitudes 0.948128 e 0.869780, respectivamente) com a variável dependente PE, mas inversamente proporcionais (sinal negativo).

Por outro lado, RH e AP apresentam correlações moderada e forte (magnitudes 0.389794 e 0.518429), ambas diretamente proporcionais (sinal positivo).
<hr>

2. Calcule a correlacão entre as variáveis independentes (AT, V, AP, RH) entre si. Existe alguma relacão forte entre elas? Se sim, isso pode indicar multicolinearidade.

In [2]:
print(x.corr())

          AT         V        AP        RH
AT  1.000000  0.844107 -0.507549 -0.542535
V   0.844107  1.000000 -0.413502 -0.312187
AP -0.507549 -0.413502  1.000000  0.099574
RH -0.542535 -0.312187  0.099574  1.000000


As variáveis independentes AT e V apresentam uma correlação muito forte (de magnitude 0.844107) e diretamente proporcional (sinal positivo).
<hr>

3. Construa um modelo de regressão linear múltipla utilizando as variáveis independentes (AT, V, AP, RH)
para prever a variável dependente (PE). Divida os dados em treino (80%) e teste (20%).

In [3]:
import pandas as pd

df = pd.read_csv("datasets/Power Plant Data.csv")

print(df.head())

X = df.drop(columns=["PE"])
y = df["PE"]

from sklearn.preprocessing import MinMaxScaler

scaler  = MinMaxScaler(feature_range=(0, 1))
scaled  = scaler.fit_transform(X)
df_scaled = pd.DataFrame(scaled, columns=X.columns)
df_scaled

from sklearn.model_selection import train_test_split

X = df_scaled
X_train, X_test, y_train, y_test = train_test_split(
  X, y, test_size=0.2, random_state=42
)

print(f"Tamanho do dataset = {X.shape[0]} amostras")
print(f"Quantidade de amostras de treinamento = {X_train.shape[0]} amostras")
print(f"Quantidade de amostras de teste = {X_test.shape[0]} amostras")

from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

theta = model.coef_
print(theta)

theta_0 = model.intercept_
print(theta_0)

      AT      V       AP     RH      PE
0  14.96  41.76  1024.07  73.17  463.26
1  25.18  62.96  1020.04  59.08  444.37
2   5.11  39.40  1012.16  92.14  488.56
3  20.86  57.32  1010.24  76.64  446.48
4  10.82  37.50  1009.23  96.62  473.90
Tamanho do dataset = 9568 amostras
Quantidade de amostras de treinamento = 7654 amostras
Quantidade de amostras de teste = 1914 amostras
[-70.10225923 -13.04365905   2.51349834 -11.7955869 ]
502.804920338107


Equação gerada pela regressão linear:

$$\begin{aligned}
\hat{y} = 502.80 & -70.10x_1 - 13.04x_2 + 2.51x_3 -11.79x_4
\end{aligned}$$
<hr>

4. Avalie o desempenho do modelo utilizando métricas como R², RMSE e MAE. O modelo é adequado para
prever a producão de energia? Justifique sua resposta.

In [4]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
import numpy as np

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, y_pred)


print(f"Erro quadrático médio (MSE): {mse:.2f}")
print(f"Coeficiente de determinação (R²): {r2:.2f}")
print(f"Raiz do Erro quadrático médio (RMSE): {rmse:.2f}")
print(f"Erro absoluto médio (MAE): {mae:.2f}")

Erro quadrático médio (MSE): 20.27
Coeficiente de determinação (R²): 0.93
Raiz do Erro quadrático médio (RMSE): 4.50
Erro absoluto médio (MAE): 3.60


Com base no $R^2$ obtido, o modelo encontrado é interessante de ser utilizado, pois de acordo com a métrica, ele explica 93% da variabilidade dos dados. O MAE possuindo um erro de 0.8% e o RMSE um erro de 1% também corroboram para o fato de que este modelo previsto é adequado para prever a produção de energia, visto que métricas de erros baixas mitigam a possibilidade de valores irreais de serem apresentados durante o treinamento.
<hr>